patroni/dcs/__init__.py

```diff
@@ -85,38 +87,78 @@ def dcs_modules() -> List[str]:
-        except ImportError:
-            logger.info('Failed to import %s', module_name)
-    raise PatroniFatalException("""Can not find suitable configuration of distributed configuration store
-Available implementations: """ + ', '.join(sorted(set(available_implementations))))
+
+def find_dcs_class_in_module(module: ModuleType) -> Optional[Type['AbstractDCS']]:
+    """Try to find the implementation of :class:`AbstractDCS` interface in *module* matching the *module* name.
+
+    :param module: Imported DCS module.
+
+    :returns: class with a name matching the name of *module* that implements :class:`AbstractDCS` or ``None`` if not
+              found.
+    """
+    module_name = module.__name__.rpartition('.')[2]
+    return next(
+        (obj for obj_name, obj in module.__dict__.items()
+         if (obj_name.lower() == module_name
+             and inspect.isclass(obj) and issubclass(obj, AbstractDCS))),
+        None)
+
+
+def get_dcs(config: Union['Config', Dict[str, Any]]) -> 'AbstractDCS':
+    """Attempt to load a Distributed Configuration Store from known available implementations.
+
+    .. note::
+        Using the list of available DCS modules returned by :func:`iter_dcs_modules` attempt to dynamically import and
+        instantiate the class that implements a DCS using the abstract class :class:`AbstractDCS`.
+
+        Basic top-level configuration parameters retrieved from *config* are propagated to the DCS specific config
+        before being passed to the module DCS class.
+
+        If no module is found to satisfy configuration then report and log an error. This will cause Patroni to exit.
+
+    :raises :exc:`PatroniFatalException`: if a load of all available DCS modules have been tried and none succeeded.
+
+    :param config: object or dictionary with Patroni configuration. This is normally a representation of the main
+                   Patroni
+
+    :returns: The first successfully loaded DCS module which is an implementation of :class:`AbstractDCS`.
+    """
+    for name, module in iter_dcs_modules(config):
+        dcs_class = find_dcs_class_in_module(module)
+        if dcs_class:
+            # Propagate some parameters from top level of config if defined to the DCS specific config section.
+            config[name].update({
+                p: config[p] for p in ('namespace', 'name', 'scope', 'loop_wait',
+                                       'patronictl', 'ttl', 'retry_timeout')
+                if p in config})
+            # From citus section we only need "group" parameter, but will propagate everything just in case.
+            if isinstance(config.get('citus'), dict):
+                config[name].update(config['citus'])
+            return dcs_class(config[name])
+
+    raise PatroniFatalException(
+        f"Can not find suitable configuration of distributed configuration store\n"
+        f"Available implementations: {', '.join(sorted(set(name for name, _ in iter_dcs_modules())))}")
```

Comment:
The original implementation was also checking that the module has a class implementing `AbstractDCS`: https://github.com/zalando/patroni/pull/2747/files#diff-247edd60473c0933e528a1b4621cc612da54b6dc964720719d7623135d312474L114-L115

I would suggest to move this filter (`find_dcs_class_in_module`) inside `iter_dcs_modules()`, which actually could just return DCS classes.

Reply:
Pushed a commit, let me know if this is what you were thinking?